### PR TITLE
1.12

### PR DESCRIPTION
**Pull request by rom1504**
can connect.

I want to do if(version == "1.12") like in mineflayer because I don't think it's complicated actually.

To do : 
* [x] fix tests : do multiple version tests like nmp and mineflayer
* [ ] check everything
* [x] positions
* [x] send unload chunk in 1.12
* [x] do place block properly (using inventory and not heldItem packet property)
* [x] check https://github.com/PrismarineJS/flying-squid/pull/310 https://github.com/PrismarineJS/flying-squid/pull/286 and https://github.com/PrismarineJS/flying-squid/issues/219
* [x] check http://wiki.vg/Protocol_History and http://wiki.vg/Protocol vs http://wiki.vg/index.php?title=Protocol&oldid=7368

### CODE REVIEW

**Comment by mhsjlw** [comment]
FYI when I took a go at this there was an issue with placing or breaking blocks. Some packet no longer has some `hand` field I believe (that gives the item in hand)

**Comment by rom1504** [comment]
We should be concerned by : 
* [ ] Removed "Leash" boolean from Attach Entity - Attach entity is now used only for leashing and Set 
* [ ] Added Set Passengers (0x3A)
* [ ] Added Teleport ID VarInt field to Player Position and Look (0x2E - clientbound)
* [ ] Added Teleport Confirm (0x1D - serverbound) packet, with a single VarInt for the teleport ID
* [ ] Added Sound Event (numeric ID) with ID 0x48. Structure is VarInt for the ID; x, y, and z as ints; volume as float; and pitch as a byte.
* [ ] Added a UUID to Spawn Painting (0x04) after entity ID
* [ ] Added VarInt enum category to Named Sound Effect (0x19)
* [ ] Added VarInt enum category to Sound Effect (0x47)
* [ ] Changed fixed-point ints for x, y, and z coordinates to doubles in Spawn Object (0x00), Spawn Experience Orb (0x01), Spawn Global Entity (0x02), Spawn Mob (0x03), Spawn Player (0x05), and Entity Teleport (0x4A).
* [ ] Changed byte-based fixed deltas measured in 32ndths of a block to short-based fixed deltas measured as (currentX * 32 - prevX * 32) * 128 in Entity Relative Move (0x25) and Entity Look And Relative Move (0x26). As such, these packets are used for movements of at max 8 blocks instead of at max 4 blocks.
* [ ] Changed cursor position values in Player Block Placement (0x1C) from ints ranging from 0 to 15 to floats ranging from 0 to 1.

**Comment by rom1504** [comment]
* [ ] all spawn packet have now uuid : what to put for object and mobs ?

**Comment by rom1504** [comment]
priorities : 
* [ ] positions : 
  * [ ] updatePositions.js
  * [ ] particle.js
  * [ ] entities.js
  * [ ] spawn.js
  * [ ] tp.js
* [ ] delta
  * [ ] physics.js

**Comment by rom1504** [comment]
I'm going to store all the positions as double (new system), and convert back to fixed point for 1.8

**Comment by rom1504** [comment]
Did the position thing.

I'm pretty happy of having finally figured out the solution to store position as double without loss of precision.

See the comment in updatePosition.js for more info.

The remaining things to do here are small stuff.

File: test/simple.test.js [hunk 1]
```diff
@@ -5,33 +5,46 @@ const squid = require('flying-squid')
...
-
-    serv.on('listening', () => {
-      done()
+const { firstVersion, lastVersion } = require('./common/parallel')
+
+squid.supportedVersions.forEach((supportedVersion, i) => {
+  if (!(i >= firstVersion && i <= lastVersion)) {
+    return
+  }
+  const PORT = Math.round(30000 + Math.random() * 20000)
```
Comment (mhsjlw) [inline test/simple.test.js]: This doesn't smell good

Comment (rom1504) [reply test/simple.test.js]: why ?
This is what I do in mineflayer and node-minecraft-protocol. It allows for running multiple tests at the same time without any port collision

**Comment by rom1504** [comment]
Tests adapted.
Mineflayer test 1.12 is not passing because of something about prismarine-chunk. See https://circleci.com/gh/PrismarineJS/flying-squid/820#queue-placeholder/containers/1

File: test/simple.test.js (hunk 1 above)
Comment (mhsjlw) [reply test/simple.test.js]: You should use the `CIRCLE_NODE_INDEX` environment variable instead, then add that to `3000`, for example.

Comment (rom1504) [reply test/simple.test.js]: Ah yeah I could do that. It's unlikely to be a problem but why not. I'll change that

Comment (rom1504) [reply test/simple.test.js]: No, that won't work locally.

Comment (mhsjlw) [reply test/simple.test.js]: `process.env.CIRCLE_NODE_INDEX || 0`

Comment (mhsjlw) [reply test/simple.test.js]: Or even `|| (Math.random() * 20000)`

Comment (rom1504) [reply test/simple.test.js]: fixed, going full deterministic

Comment (rom1504) [reply test/simple.test.js]: used port 0. Much more reliable. Thanks for insisting on that issue !

**Comment by rom1504** [comment]
left : 
* [ ] attach
* [ ] checking stuff
* [ ] does saving still work ?

**Comment by mhsjlw** [comment]
This features file is really cool, perhaps we should expand it to be admin configuration? So that they can enable/disable features?

**Comment by rom1504** [comment]
Maybe it could use the same pattern but I'd like to keep this one for minecraft version support for now. It's still a bit new and I'm not sure what are all the consequences.
I think there should be a vanillaFeature.json (which is what features.json is currently), and then we can possibly have more feature files if it makes sense for other use cases.

**Comment by rom1504** [comment]
sounds are not converted for 1.12 but let's not do it here (it needs more data and stuff), tracked there https://github.com/PrismarineJS/flying-squid/issues/340

**Comment by rom1504** [comment]
works nicely. Let's merge

**Pull request merged**